app/controllers/application_controller.rb

```diff
@@ -1,10 +1,154 @@
 require './config/environment'
+require 'pry'
 
 class ApplicationController < Sinatra::Base
 
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "encription_key_that_you_wont_guess"
   end
 
-end
\ No newline at end of file
+  get '/' do
+    @user = User.find_by_id(session[:user_id])
+    erb :index
+  end
+
+  get '/signup' do
+    if logged_in?
+      redirect to '/tweets'
+    else
+      erb :'users/create_user'
+    end
+  end
+
+  post '/signup' do
+    if params[:username] == "" || params[:email] == "" || params[:password] == ""
+      redirect to '/signup'  #how can you raise an error here? "You must fill in the all parts of the form to create an account."
+    else
+      @user = User.create(username: params[:username], email: params[:email], password: params[:password])
+      @user.save
+      session[:user_id] = @user.id
+      redirect to '/tweets'
+    end
+  end
+
+  get '/login' do
+    if logged_in?
+      redirect to '/tweets'
+    else
+      erb :'users/login'
+    end
+  end
+
+  post '/login' do
+    @user = User.find_by(username: params[:username])
+    if @user && @user.authenticate(params[:password])
+      session[:user_id] = @user.id
+      redirect '/tweets'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/logout' do
+    if logged_in?
+      session.clear
+      redirect to '/login'
+    else
+      redirect to '/'
+    end
+  end
+
+  get '/users/:slug' do  #get "/users/#{user.slug}" shows all a single users tweets'
+    @user = User.find_by_slug(params[:slug])
+    erb :'users/show'
+  end
+
+  get '/tweets' do
+    if logged_in?
+      @user = User.find_by_id(session[:user_id])
+      @tweets = Tweet.all
+      erb :'tweets/tweets'
+    else
+      redirect to '/login'
+    end
+  end
+
+  get '/tweets/new' do
+    if logged_in?
+      erb :'tweets/create_tweet'
+    else
+      redirect to '/login'
+    end
+  end
+
+  post '/tweets' do
+    if params[:content] != ""
+      @tweet = Tweet.create(content: params[:content], user_id: session[:user_id])
+      redirect to  "/tweets/#{@tweet.id}"
+    else
+      redirect to '/tweets/new'
+    end
+  end
+
+  get '/tweets/:id' do
+    if logged_in?
+      @tweet = Tweet.find_by_id(params[:id])
+      erb :'tweets/show_tweet'
+    else
+      redirect to '/login'
+    end
+  end
+
+  get '/tweets/:id/edit' do
+    if logged_in?
+      @tweet = Tweet.find_by_id(params[:id])
+      erb :'/tweets/edit_tweet'
+    else
+      redirect to '/login'
+    end
+  end
+
+  patch '/tweets/:id' do
+    if logged_in?
+      @tweet = Tweet.find_by_id(params[:id])
+      if params[:content] == ""
+        redirect to "/tweets/#{@tweet.id}/edit"
+      else
+        @tweet.content = params[:content]
```

Comment:
I would contend that you still want to double check the user's association to the tweet here. There are other ways to submit a PATCH/POST request than going through an HTML form...